### PR TITLE
Sync Discord and Scrimba avatars

### LINKED ISSUE #29: Monitor avatar changes
Can we listen for changes to a discord users avatar and update the users scrimba avatar when there is a change?

That would be to update the users `avatar_image` with a new url provided by discord. The url should look like this one: https://cdn.discordapp.com/avatars/500607885714128897/9009212b5d0bcf3e3af2505cf140e6ec.png

### PR DESCRIPTION
**Pull request by codeluggage**
Closes #29 

## Changes

- Listens to `guildMemberUpdate` for server-local changes to avatars and refreshes the connected Scrimba avatar if there is a change.
- Listens to `presenceUpdate` and refreshes the connected Scrimba avatar if there is a change.
- Includes a few general improvements to types that are outside of the scope of this PR, like the concept of log levels throughout the bot codebase, and moving towards a class based architecture with dependency injection.

## Open questions

- [x] How are we organizing code from an architecture point of view, now that it's all Imba? Early intent here is to move towards classes and composition. 
- [] ~~Would it be fun for the community to have an API to update their avatar through? Would open for challenges in the future, updating their avatar programatically.~~

## Deploy

- [x] Add "Presence Intent" permission to the Application dashboard for the bot, so it knows when users log into the server

### CODE REVIEW

File: src/ScrimbaIntegration.imba [hunk 1]
```diff
@@ -0,0 +1,39 @@
+import { Client } from "discord.js"
+import type { Knex } from 'knex'
+import type { GuildMember, TextChannel, GuildChannel } from "discord.js"
+# import {config} from "../config"
+import { log_discord, log_error, log_info } from "./utils/logger"
```
Comment (Pistus) [inline src/ScrimbaIntegration.imba:5]: I think this would be better as a named import so that the calls are log.info, log.error, etc.

**Review by Pistus — COMMENTED** [review]
Had a quick first look. In general looks like a good direction.

**Comment by Pistus** [comment]
`How are we organizing code from an architecture point of view, now that it's all Imba? Early intent here is to move towards classes and composition.`

- Ok, sounds like a good start

`Would it be fun for the community to have an API to update their avatar through? Would open for challenges in the future, updating their avatar programatically.`

- Do you then mean that they can upload images or paste any link to be their avatar? This doesn't sound like great value to me.

**Comment by codeluggage** [comment]
> `Would it be fun for the community to have an API to update their avatar through? Would open for challenges in the future, updating their avatar programatically.`
> 
>     * Do you then mean that they can upload images or paste any link to be their avatar? This doesn't sound like great value to me.

No, rather that it could be a "learn a little backend" part of a course. Specifically, being able to see their avatar change in the navbar as they run code. 

In any case, we're moving away from this approach and there is a simpler implementation I'm pushing up later, so we can skip this thought 👍

File: src/ScrimbaIntegration.imba [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+export class ScrimbaIntegration
+	prop knex\Knex
+
+	def constructor(client\Client, knex\Knex)
+		knex = knex
+		client.on "guildMemberUpdate", onGuildMemberUpdate.bind(self)
+		client.on "presenceUpdate", onPresenceUpdate.bind(self)
+
+	def avatar_needs_refresh\Promise<string | undefined> user\User, member\GuildMember?
```
Comment (Pistus) [inline src/ScrimbaIntegration.imba:13]: I got two suggestions for the naming convention here. Let's see if any of them fit:)

Number 1: A predicate ala. `refreshAvatar?`

Number 2: The camelCase version as that is the defacto standard in the repo ala. `avatarNeedsRefresh`

File: src/ScrimbaIntegration.imba [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+		return unless const new_avatar = (member..avatarURL! or user.avatarURL!)
+		const old_avatar = await knex.select('avatar_image').from('users').where(discord_id: user.id)
+		unless old_avatar.length > 0
+			# TODO: Turn this on to help us understand how many are "pre onboarding" and don't have connected accounts
+			# log.i "No scrimba user found for discord_id: { user.id }"
+			return
+
+		return new_avatar if new_avatar != old_avatar[0].avatar_image
+
+	def refresh_avatar\Promise<void> user\User, new_avatar\string
```
Comment (Pistus) [inline src/ScrimbaIntegration.imba:24]: `update_avatar` seems more precise and consistent with discord.js + scrimba-site repo. Or `updateAvatar` if you get on the camel ride 👆

**Review by Pistus — CHANGES_REQUESTED** [review]
Some hygiene comments - the main syncing code looks good. I have failed to complete the setup of my own discord server so haven't tested it yet (have copied a template of the scrimba server, but stopped there). But I am happy to ship this as it is quite contained and the commit with the logic was easy to go through.

File: src/utils/logger.imba [hunk 4]
```diff
@@ -0,0 +1,47 @@
+import type { Guild } from 'discord.js'
+import { MessageEmbed, TextChannel } from 'discord.js'
+
+const PUMPKIN_LOG = 'pumpkin-log' # TODO: get channel from config env
+
+def to_channel\Promise<void> channel_source\Guild, type\string, description\string, image\string?
```
[the remark this answers is not in the document]
Comment (codeluggage) [reply src/utils/logger.imba:6]: +1 - I'd love to do this. I have gleaned quite a bit of best practices around this from the recent investigation into posting for the users, and wrapping a class around the `Client` would set us up for simplified logging. That would also make the code more terse, as the initial creation of the logger can take in what the `type` would be, etc. 

Deferring this until we refactor the functional `onboarding.imba` and `karma.imba` files.

**Review by Pistus — APPROVED** [review]
Ok, logger issue to be handled later - looks good.

**Pull request merged**